go/arrow/array/half_float_builder.go

```diff
@@ -0,0 +1,155 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package array
+
+import (
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/internal/bitutil"
+	"github.com/apache/arrow/go/arrow/internal/debug"
+	"github.com/apache/arrow/go/arrow/memory"
+	"sync/atomic"
+)
+
+type HalfFloatBuilder struct {
+	builder
+
+	data    *memory.Buffer
+	rawData []arrow.Float16
+}
+
+func NewHalfFloatBuilder(mem memory.Allocator) *HalfFloatBuilder {
+	return &HalfFloatBuilder{builder: builder{refCount: 1, mem: mem}}
+}
+
+// Release decreases the reference count by 1.
+// When the reference count goes to zero, the memory is freed.
+func (b *HalfFloatBuilder) Release() {
+	debug.Assert(atomic.LoadInt64(&b.refCount) > 0, "too many releases")
+
+	if atomic.AddInt64(&b.refCount, -1) == 0 {
+		if b.nullBitmap != nil {
+			b.nullBitmap.Release()
+			b.nullBitmap = nil
+		}
+		if b.data != nil {
+			b.data.Release()
+			b.data = nil
+			b.rawData = nil
+		}
+	}
+}
+
+func (b *HalfFloatBuilder) Append(v float32) {
```

Comment:
here, I am on the fence: should we accept a `float32` or a `float16`?
it looks nicer in tests (where we mostly deal with literal values) but in real life situations, we usually deal with variables that already have the correct type (_ie:_ `float16`) so this would incur an explicit conversion to `float32` via `v.Float32()` (and immediately convert back to `float16`.)
this sounds a bit wasteful.

but I am not strongly opposed to using `float32` in the API.

Reply:
same here, we can provide one method for float32, and another for float16.